### PR TITLE
Update RM implementation to account for talent procs

### PR DESCRIPTION
**Pull request by trevorm4**
Rising Mist module currently blindly extends all HoTs but this is not reality. ReM and EnvM from talents (Dancing Mists, Misty Peaks, and Rapid Diffusion) do not get extended by ReM. This diff adds a module to link Buff applications/removal to their original application. Iff a ReM can be linked back to a hardcast then it will be extended by ReM. This framework also adds support for adding attributions for various talents, tho atm only hardcast is added to the hot object.

### CODE REVIEW

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Are bounces able to cause a RefreshBuff?

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Well this is novel, but probably works. How was the 65 second buffer chosen?

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Big possibility of false positive here - I'm pretty sure if player casts Enveloping Mists -> Renewing Mist, both spells will go off at the same time, and then this link will catch the hardcast.

While it's not documented, the eventlinks are applied in order so I think you could add an additional condition to this link to check that it doesn't already have a FROM_HARDCAST link. You'd be breaking new ground for this normalizer, but it was a use case I had in mind.

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Again I'd think about the possibility of false positives here although I haven't thought this all the way through

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Big yikes, but I think I understand what you're doing here. Looks like you're resolving the 'double link' problem with logic here, then tracing back bounces if you need to? This seems viable, but I'd like you to add a bit more docs just so future maintainers can figure out the reasoning.

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Does the buffer really need to be this big, and in both directions?

**Review by kfinch — CHANGES_REQUESTED** [review]
Looks good, but I'd like you to take a second look on the things I've commented on

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (trevorm4) [reply src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Talked to Sweggles and he said 77k ms is a good safe upper bound as you can get a bunch of extensions from casting EF and FLS EF on cd, so gonna change to that. With just baseline tft rem and max RM extension the max is 60 seconds, but each EF bolt hit adds another second.

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (trevorm4) [reply src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Yep technically if every single person in the group has a rem then yes, should add RefreshBuffEvent alongside Apply's.

**Comment by kfinch** [comment]
I discussed strategies for this in DM convo with Trevor - brainstorming solutions

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (trevorm4) [reply src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Addressing this one by adding an additionalCondition that it cannot be linked to a hardcast as suggested in DMs, which should address false positive issue

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (trevorm4) [reply src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Added comments and split out spooky logic to a helper function to explain this one as it is a very necessary step.

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (trevorm4) [reply src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts]: Reducing buffer

File: src/analysis/retail/monk/mistweaver/modules/spells/RisingMist.js
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/modules/spells/RisingMist.js:142]: Can do same thing with `return Boolean(attributions.find(att => att.name === name));`

File: src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts
Comment (kfinch) [inline src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts:183]: This is pretty scary, but if it's working I guess??

**Comment by kfinch** [comment]
After discussion, will merge this now and look for avenues for improvement in follow-on PR

**Review by kfinch — APPROVED** [review]

**Pull request merged**
